STUDIO/org.openl.rules.workspace/src/org/openl/rules/project/abstraction/ZipFolderRepository.java

```diff
@@ -131,10 +145,40 @@ public Features supports() {
     }
 
     private ZipInputStream getZipInputStream() throws IOException {
-        FileItem fileItem = version == null ? delegate.read(zipPath) : delegate.readHistory(zipPath, version);
+        FileItem fileItem = getFileItemWithRetry();
         return new ZipInputStream(fileItem.getStream());
     }
 
+    private FileItem getFileItemWithRetry() throws IOException {
+        FileItem fileItem = getFileItem();
+        if (fileItem == null) {
+            for (int i = 0; i <= GET_FILE_RETRY_ATTEMPTS; i++) {
+                ScheduledExecutorService scheduledExecutorService = Executors.newScheduledThreadPool(1);
+                Callable<FileItem> getFileItemTask = this::getFileItem;
+                ScheduledFuture<FileItem> schedule = scheduledExecutorService.schedule(getFileItemTask, GET_FILE_RETRY_DELAY_S, TimeUnit.SECONDS);

```

Comment:
Bad solution to retry in the separate thread in this special class. OpenL RuleServices already has retry functionality on the upper layer in the Deployment manager. According to the API, Repository#read() function can return null and it is a valid case. It needs just to handle this case correctly, when no data is read from the repository. Another question, how to prevent null at all in this case, so fix must be in the Deployment manager.